### PR TITLE
8204681: Option to include timestamp in hprof filename

### PR DESCRIPTION
**Pull request by SoniaZaldana**
Hi all, 

This PR addresses [8204681](https://bugs.openjdk.org/browse/JDK-8204681) enabling support for timestamp expansion in filenames specified in `-XX:HeapDumpPath` using `%t`. 

As mentioned in this comments for this issue, this is somewhat related to [8334492](https://bugs.openjdk.org/browse/JDK-8334492) where we enabled support for `%p` for filenames specified in jcmd. 

With this patch, I propose: 
- Expanding the utility function `Arguments::copy_expand_pid` to `Arguments::copy_expand_arguments` to deal with `%p` expansions for pid and `%t` expansions for timestamps. 
- Leveraging the above utility function to enable argument expansion for both heap dump filenames and jcmd output commands. 
- Though the linked JBS issue only relates to heap dumps generated in case of OOM, I think we can edit it to more broadly support filename expansion to support `%t` for jcmd as well. 

Testing: 
- [x] Added test cases pass with all platforms (verified with a GHA job). 
- [x] Tier 1 passes with GHA. 

Looking forward to hearing your thoughts!

Thanks, 
Sonia

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warnings
&nbsp;⚠️ Patch contains a binary file (src/java.base/share/classes/jdk/internal/icu/impl/data/icudt74b/nfc.nrm)
&nbsp;⚠️ Patch contains a binary file (src/java.base/share/classes/jdk/internal/icu/impl/data/icudt74b/nfkc.nrm)
&nbsp;⚠️ Patch contains a binary file (src/java.base/share/classes/jdk/internal/icu/impl/data/icudt74b/ubidi.icu)
&nbsp;⚠️ Patch contains a binary file (src/java.base/share/classes/jdk/internal/icu/impl/data/icudt74b/uprops.icu)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/javax/swing/text/doc-files/Document-insert.gif)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-bw16.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-bw24.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-bw32.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-bw48.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-interim16.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-interim24.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-interim32.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-interim48.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-yellow16.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-yellow24.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-yellow32.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-yellow48.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/windows/native/libawt/windows/security_warning.ico)
&nbsp;⚠️ Patch contains a binary file (src/utils/IdealGraphVisualizer/View/src/main/resources/com/sun/hotspot/igv/view/images/hideDuplicates.png)
&nbsp;⚠️ Patch contains a binary file (test/jdk/java/security/ProtectionDomain/AllPerm.jar)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyFile/TokenStore.keystore)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyFile/TrustedCert.keystore)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyFile/TrustedCert.keystore1)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyParser/ExtDirsA/a.jar)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyParser/ExtDirsB/b.jar)

### Issue
 * [JDK-8204681](https://bugs.openjdk.org/browse/JDK-8204681): Option to include timestamp in hprof filename (**Enhancement** - P3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20568/head:pull/20568` \
`$ git checkout pull/20568`

Update a local copy of the PR: \
`$ git checkout pull/20568` \
`$ git pull https://git.openjdk.org/jdk.git pull/20568/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20568`

View PR using the GUI difftool: \
`$ git pr show -t 20568`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20568.diff">https://git.openjdk.org/jdk/pull/20568.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20568#issuecomment-2286976265)
</details>

### CODE REVIEW

**Comment by SoniaZaldana** [comment]
cc @kevinjwalls

**Comment by tstuefe** [comment]
I think this could be very useful, but it needs more preparation and decisions. Possibly a CSR.

- copy_expand_xxx is used in many places. While I think all of these places would benefit from more expansions than just %p, there is a potential backward compatibility issue if clients use %t for whatever reason today
- Do we want the time of the dump or the JVM start? If the JVM runs for a week, then produces a JFR file, should the file be named by the JVM start date? I think in most cases the *current* time makes more sense
- Do we want the printout as a human-readable date or as a numeric timestamp? Both makes sense depending on the post-processing clients want to do.
- Do we want to improve this function further, potentially adding more replacement options?

One possible way to solve this:
- use different characters for timestamp (number) and datetime (human readable date)
- use always the current time
- If we want to add further replacements: 
  - come up with a new replacement character that does not clash with libc sprintf (IMHO using percent was not a good idea in the first place). E.g. `$`
  - Add a new switch to guard this new replacement logic. By default off. If on, the contract is that any character following a `$` may be either now or in the future replaced with something different. Client must not use `$` as a normal character.
  - We probably should remove all non-matching `$` from the input.
  - The first replacements could be: `$p` for pid, `$t` for timestamp (numeric), `$d` for datetime
  - later replacements can be added later. Since we guard the new feature with a switch and forbid the use of `$`, we are then free to do so without breaking backward compatibility.

**Review by kevinjwalls — COMMENTED** [review]
Hi,
Yes, agreed with notes above that current time is the timestamp that is useful.
(Multiple files from the same VM are created with the same filename pattern, and file created is different over time.  Good.  And if the pid is the same, timestamp keeps things "more" unique.)
That means we should not need any changes relating to start_time = create_vm_timer...? ( src/hotspot/share/runtime/threads.cpp )

I'd imagined a raw numerical timestamp, to keep things simple, the example in JDK-8204681 was like that.
However src/hotspot/share/utilities/ostream.cpp has make_log_name() which says "%t => YYYY-MM-DD_HH-MM-SS" so maybe we could follow that pattern.  As long as it's clear, and we'll be doing help/man page updates also.

JDK-8204681 was logged originally about -XX:HeapDumpPath but yes all of our output filenames from diagnostic commands would ideally support this.

FYI, in JDK-8338603 I will try and clear up the "FILE" type confusion.  It is not easy for a JMX client to know that implementation-dependent types may exist.

I can come back for a proper review soon.

**Comment by dcubed-ojdk** [comment]
> which says "%t => YYYY-MM-DD_HH-MM-SS" so maybe we could follow that pattern

Another vote for the above time stamp pattern.

**Comment by SoniaZaldana** [comment]
Hi folks, thanks for the intial comments. Just noting I will be away on holiday until September 9th, so I will not be able to address these until then.

**Comment by dholmes-ora** [comment]
I don't object (don't really have strong views) on adding this functionality, but as @tstuefe notes there are a few things to consider. I'm not really averse to using the `%` character precisely because it is commonly identified as a format specifier - and I think `$` would be very problematic due to shell issues. At the risk of seeking the perfect instead of just doing what is immediately "good enough" we might also look at the unified logging decorators as potential formats:
```
Available log decorators:
 time (t), utctime (utc), uptime (u), timemillis (tm), uptimemillis (um), timenanos (tn), uptimenanos (un), hostname (hn), pid (p), tid (ti), level (l), tags (tg)
```
and it may also allow for some code sharing.

File: test/hotspot/jtreg/runtime/ErrorHandling/TestHeapDumpFilenameExpansion.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    private static void testTimestampSusbtitution(String filename) throws Exception {
+        Process process = startProcess(filename);
+        OutputAnalyzer output = new OutputAnalyzer(process);
+        output.stdoutShouldNotBeEmpty();
+
+        boolean found = false;
+        Pattern pattern = Pattern.compile("file\\d{4}-\\d{2}-\\d{2}_\\d{2}-\\d{2}-\\d{2}");
+        File[] files = new File(".").listFiles();
+        if(files != null) {
```
Comment (turbanoff) [inline test/hotspot/jtreg/runtime/ErrorHandling/TestHeapDumpFilenameExpansion.java:90]:
```suggestion
        if (files != null) {
```

File: test/jdk/sun/tools/jcmd/TestJcmdArgumentSubstitution.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    private static void verifyFilenameTimestampSubstitution(String... args) throws Exception {
+        String test_dir = System.getProperty("test.dir", ".");
+        OutputAnalyzer output = JcmdBase.jcmd(args);
+        output.shouldHaveExitValue(0);
+
+        boolean found = false;
+        Pattern pattern = Pattern.compile("myfile\\d{4}-\\d{2}-\\d{2}_\\d{2}-\\d{2}-\\d{2}");
+        File[] files = new File(test_dir).listFiles();
+        if(files != null) {
```
Comment (turbanoff) [inline test/jdk/sun/tools/jcmd/TestJcmdArgumentSubstitution.java:87]:
```suggestion
        if (files != null) {
```

**Comment by SoniaZaldana** [comment]
Hi folks, 

Sorry about the late reply. Just got back in from PTO. 

I’m glad to see that there is some consensus that diagnostic commands should ideally support timestamp expansion for output filenames (with the possibility of adding more replacement options in the future). 

Thomas raised the issue about backwards compatibility issues with `Arguments::copy_expand_arguments`. We’ve been discussing some ways to address this: 

1. Introduce a new parameter for jcmd that supports filename expansion (`-filepattern` or something along those lines.) We would then need to sort out ignoring `-filename` if a filename pattern is provided. 
2. Replacing `%t` with a pattern that is very unlikely to have been used and expanded by accident (`%TIMESTAMP`, etc). 
3. A HotSpot flag (`-XX:+AllowFileNamePattern`) to accept these patterns in multiple places in the JVM. 

Option 2 is definitely the most straightforward to implement but it’s also not the cleanest. There is always a slight risk that someone’s configuration might break. 

I’d really like to find an actionable consensus on what would be best. Open to explore other options if anyone has any better ideas. 

Aside from that, I have noted the feedback about using current time vs start time and the `YYYY-MM-DD_HH-MM-SS` format. I will be making the appropriate changes once we’ve settled the other stuff. 

Thanks!

**Comment by tstuefe** [comment]
> I don't object (don't really have strong views) on adding this functionality, but as @tstuefe notes there are a few things to consider. I'm not really averse to using the `%` character precisely because it is commonly identified as a format specifier - and I think `$` would be very problematic due to shell issues. At the risk of seeking the perfect instead of just doing what is immediately "good enough" we might also look at the unified logging decorators as potential formats:
> 
> ```
> Available log decorators:
>  time (t), utctime (utc), uptime (u), timemillis (tm), uptimemillis (um), timenanos (tn), uptimenanos (un), hostname (hn), pid (p), tid (ti), level (l), tags (tg)
> ```
> 
> and it may also allow for some code sharing.

This is a very good thought. There were people that already thought about useful things to have in a log. This is very similar.

**Comment by kevinjwalls** [comment]
A few notes, without reviewing the code further today, see if these all make sense and are settled:

"%t" expanded to timestamp seems simple enough (I think % is not controversial, extending from our existing use of %p)


But yes what is a timestamp?  There could be debate on the format.
The original issue came from a user request, but it was not fully specified, just a request for an automatic way of putting file creation time in an output filename.

So should we have %t and %d ? Or adopt all the decorator options from unified logging?  To me those seem extreme: if I just want output files to say when they were made, I don't really need options, or nanoseconds, and would be quite happy with the ostream.cpp style of "%t => YYYY-MM-DD_HH-MM-SS".


Sonia your points above:
(I haven't looked in detail at how this affects copy_expand_arguments, but if we can settle the jcmd options then that can follow...)

1. -filepattern
Would really like to avoid adding a new jcmd option, in all the diagnostic commands that take an output filename, and just go with a filename that respects some substitutions.

2
%t and %p are "unique enough" that we don't need "%TIMESTAMP" ?
We don't need to permit creating "my%perfect%test%file" without using "%%" to get a literal % included?

3.
-XX:+AllowFileNamePattern 
I would think at this step, we get our jcmds that take an output FILE to respect the %t timestamp.  There may be other XX options that take filenames which could be considered in future.

**Comment by tstuefe** [comment]
> So should we have %t and %d ? Or adopt all the decorator options from unified logging? To me those seem extreme: if I just want output files to say when they were made, I don't really need options, or nanoseconds, and would be quite happy with the ostream.cpp style of "%t => YYYY-MM-DD_HH-MM-SS".

A subset of the decorators of UL would make perfect sense, especially if we talk about a (possibly future) generic way to enrich file names:
- time (t),  utctime (UTC)  make obviously sense
- uptime (u), ... possibly (eg as a primitive way to avoid duplication for multiple dumped files in one run)
- hostname makes a lot of sense for distributed systems
- pid obvious
- tid possibly, if one has the need to dump per-thread files

I would argue for reusing these specifiers, and (either now or in the future) possibly also the code behind them, for decorating file names. Better than having to come up today with a %t, tomorrow someone maybe needs the hostname too, so %h? and so on.

**Comment by kevinjwalls** [comment]
Thanks Thomas -
Good to get this going again 8-)

Not really against other substitutions being available if somebody is going to implement them. 8-)
(and explore the possible code sharing)

Timestamp is valuable as it's a commonly needed and hard to know ahead of time.
It seems useful that we do it in the JVM, as while I can call out to `date +%...` in a terminal or in a script, I have to think a fair amount (and generate a date string in a consistent way, and without spaces and slashes).


Can't see a problem with %h hostname being added later if there is a perceived need, if this change goes ahead just doing timetamps.

I was thinking about the changes being possible incrementally, but much seems to depend on how much work the author wants to take on at once.


For parsing a filename in a command, one-letter identifiers like %p, %t, and maybe %h etc... are unambiguous to parse.

Using the unified logging decorator set, we start using two characters or more. Things get ambiguous if we don't terminate the parsing somehow, and the user has to say something like "filename%(t)issue123.out" if they don't separate things with non alphabetic characters, e.g. "filename_%t_issue123.out".  I might prefer the latter, but not sure I can say everybody has to do it that way...



Now I reread the original issue:
8204681: Option to include timestamp in hprof filename
..so that was raised about -XX:HeapDumpOnOutOfMemoryError etc...

If this change is to add %t timestamp (or more) in jcmd output files, it should all be in a new issue like "Option to include timestamp in jcmd output filenames", and in future -XX:HeapDumpOnOutOfMemoryError etc could still be expanded to respect a timestamp etc... with 8204681.

**Comment by SoniaZaldana** [comment]
Hi folks, 

Thanks for your comments. I spent some time today prototyping how to reuse UL tag logic. Like @kevinjwalls mentioned, it becomes a bit tricky when the identifiers are more than one letter. Especially as it might lead to some unintended consequences. For example, a user can specify the file name `my%tinstance` and end up with a thread id (`ti` per UL tags) versus an intended timestamp. 

I didn’t want to spend too much time on making the 2-letter identifiers work but I made some small changes to use UL tags in general. Looks like [this](https://github.com/SoniaZaldana/jdk/commit/36cdef1a598761e5310da4e30c39e818f2ff1230).

Note the timestamp is diplayed in the following format: `2024-10-31T15:14:39.344-0400`

As far as the scope of this PR, I'm okay to implement it using UL tags (one or two letters) if we feel that will serve argument expansion in the long term. However, I still have the question about backwards compatibility. Would reviewers be receptive to changing `Arguments::copy_expand_pid` to `Arguments::copy_expand_arguments` and propagate that across all places where `Arguments::copy_expand_pid` as originally proposed or is there another preferred approach? 

Thanks for taking a look! 

Sonia

**Comment by kevinjwalls** [comment]
Thanks Sonia --

Regarding timestamp format:

UL style got you a timestamp like: 2024-10-31T15:14:39.344-0400
make_log_name() would give:        YYYY-MM-DD_HH-MM-SS

Precision vs filename length?  The .001 of a second accuracy might not be that meaningful on a file name.  

I was liking the make_log_name() style above, but it's not really critical which.

But what does Unified Logging do when you log to a file?

file=filename uses %p and %t for pid and STARTUP timestamp.  Hmm.... For usage like heapdumps where you may want multiple, current time is much more useful than startup time.  But hey, it's another vote for the use of %t for timestamp in a filename, and a UnifiedLogging log to a file does not give you control over exactly how you want the timestamp in the filename.  You get what you get.  (The events logged in the file are what matter I suppose (and I agree)).

So what timestamp do you get?

java -Xlog:gc*:file%t.out -version
creates

file2024-11-01_12-01-10.out

..so that looks like the simpler version, like make_log_name()  mentioned earlier.  No milliseconds or time offsets.

I would do that, and forget multi-character patterns unless somebody really wants/needs it.  8-)


------

On changing Arguments::copy_expand_pid to Arguments::copy_expand_arguments

There are a few users, listed below.  I can't see any problem by starting to respect a new % token, but there is always a chance as I think Thomas was suggesting.


If anybody is concerned, we should keep Arguments::copy_expand_pid, and add a new function to do timestamps additionally.


------

Arguments::copy_expand_pid
I find usage by:

perfMemory.cpp:
perdata file is not normally going to have %t in it but -XX:PerfDataSaveFile= could set it.
Users would be really unwise to add specify %t here!  Maybe that means there should still be Arguments::copy_expand_pid which ONLY does %p expansion.


diagnosticArgument.cpp uses it, good, that's what we want.


vmError.cpp:
next_OnError_command uses it.
Having timestamp support in -XX:OnError command strings in addition to %p could be really useful.  Small chance of a clash with existing commands/scripts.


vmError.cpp:
expand_and_open / prepare_log_file /  VMError::report_and_die calls:  prepare_log_file(ErrorFile, "hs_err_pid%p.log",...
 
OK, being able to specify a timestamp in hs_err could be very useful.
Also the compiler replay data file would be affected (again not bad).

Also called by JVMCI::fatal_log() meaning -XX:JVMCINativeLibraryErrorFile= could set a filename using %p and now %t.



jfrEmergencyDump.cpp:
create_emergency_dump_path() uses it, with these patterns, which are not configurable, so not a problem:

44  static const char vm_error_filename_fmt[] = "hs_err_pid%p.jfr";
45  static const char vm_oom_filename_fmt[] = "hs_oom_pid%p.jfr";
46  static const char vm_soe_filename_fmt[] = "hs_soe_pid%p.jfr";


The changes look generally useful.  A release note and maybe other doc updates will be needed (which is fine).

If the impact looks too much, keeping Arguments::copy_expand_pid and creating Arguments::copy_expand_arguments which optionally does timestamp expansion, will mean the other changes (hs_err, replay, etc..) can be made separately.

**Pull request closed without merging**